server/services/discovery.py

```diff
@@ -0,0 +1,188 @@
+
+from dataclasses import dataclass
+import itertools
+from typing import Dict, List, Union
+import yaml
+
+import lib.config as libconfig
+
+cfg = libconfig.get_config()
+
+
+class InvalidEndpointException(Exception):
+  pass
+
+
+class InvalidIngressConfig(Exception):
+  pass
+
+
+# By default, all endpoints are expected on localhost port 8081.
+DEFAULT_INGRESS_RULES = dict({cfg.API_ROOT: ['/*']})
+
+
+@dataclass
+class Endpoint:
+  name: str  # Reference of the endpoint, use only within this server.
+  path: str  # Path to a specific API defined by mixer.
+
+
+class Endpoints:
+  """Container for endpoint."""
+
+  def __init__(self, endpoints: List[Endpoint]):
+    self.endpoints = endpoints
+    self.endpoint_name_to_path = {e.name: e.path for e in self.endpoints}
+    self.endpoint_name_to_host = dict()  # unconfigured.
+
+  def configure(self, ingress_rules: Dict[str, List[str]]):
+    """Sets endpoint name to host mapping from ingress rules."""
+    patterns = itertools.chain(*ingress_rules.values())
+    if '/*' not in patterns:
+      raise Exception('Invalid ingress rules: Must have a "/*" route.')
+
+    # Find a matching host for each endpoint from ingress rules.
+    for endpoint in self.endpoints:
+      host = self.find_host(endpoint.path, ingress_rules)
+      self.endpoint_name_to_host[endpoint.name] = host
+
+  def get_all_hosts(self) -> List[str]:
+    """Returns a list of all hosts."""
+    return sorted(self.endpoint_name_to_host.values())
+
+  def find_host(self, path: str, ingress_rules: Dict[str, List[str]]) -> str:
+    """Returns host with the pattern closest to a given path.
+
+    For example, suppose we have the following.
+    rules = {
+      'svc-ab': ['a/b/*'],
+      'svc-abc': ['a/b/c'],
+      'svc-b': ['/*']
+    }
+
+    Exact match overrides wildcard ending.
+    find_host('a/b/c', rules) -> 'svc-abc'
+
+    Longer prefix match takes precedence.
+    find_host('a/b/d/e', rules) -> 'svc-ab'
+
+    Wildcard match catches everything.
+    find)host('xyz/def', rules) -> 'svc-b'
+
+    For more on what is a pattern, see GCP doc.
+    https://cloud.google.com/load-balancing/docs/url-map-concepts#wildcards-regx-dynamic
+    """
+    match_host, match_pattern = '', ''
+    for host, patterns in ingress_rules.items():
+      for pattern in patterns:
+        # Pattern has wild card.
+        if pattern.endswith('/*'):
+          pattern_prefix = pattern[:-len('/*')]
+          if not path.startswith(pattern_prefix):
+            continue
+          if len(pattern) > len(match_pattern):
+            match_pattern = pattern
+            match_host = host
+
+        # Pattern does not have wild, card -> exact match.
+        if not path.startswith(pattern):
+          continue
+        if len(pattern) > len(match_pattern):
+          match_pattern = pattern
+          match_host = host
+
+    return match_host
+
+  def get_service_url(self, endpoint_name: str) -> str:
```

Comment:
In "local" development, the dc service is autopush mixer, where the url would be the mixer api url directly.

Reply:
This PR accounts for that. Assuming "local" development uses LocalBaseConfig, API_ROOT points to mixer autopush. Default ingress rule always points to that api url for all hosts.

DEFAULT_INGRESS_RULES = dict({cfg.API_ROOT: ['/*']})

Reply:
I see. Can we use real complete rules in the test and test "local", "prod" routes for a larger set of APIs?

Reply:
Added test for local (default ingress rules). Added another test that mimics prod. I think I'll add more tests are follow up, as we define what prod ingress rules we have for website's mixer.